File: cpp/subarraySum.cpp
//
// Created by Kurna on 25-7-29.
//
#include <iostream>
#include <unordered_map>
#include <vector>

class Solution {
public:
    std::unordered_map<int, int> mp;

    int subarraySum(std::vector<int> &nums, int k) {
        int sum = 0, ret = 0;
        mp[0] = 1;
        for (int x: nums) {
            sum += x;
            if (mp.find(sum - k) != mp.end()) {
                ret += mp[sum - k];
            }
            mp[sum]++;
        }
        return ret;
    }
};

int main() {
    Solution sol;
    std::vector nums = {1, 1, 1};
    auto k = 2;
    std::cout << sol.subarraySum(nums, k) << std::endl;
}

File: src/leetcode/subarraySum.py
#
# @lc app=leetcode.cn id=560 lang=python3
#
# [560] 和为 K 的子数组
#
from typing import List
# @lc code=start
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        ans = 0
        prefix_sum = 0
        prefix_sum_count = {0: 1}  # Initialize with prefix sum 0
        for num in nums:
            prefix_sum += num
            if prefix_sum - k in prefix_sum_count:
                ans += prefix_sum_count[prefix_sum - k]
            prefix_sum_count[prefix_sum] = prefix_sum_count.get(prefix_sum, 0) + 1
        return ans
            
        
# @lc code=end

def main():
    sol = Solution()
    nums = [1, 1, 1]
    k = 2
    print(sol.subarraySum(nums, k))
    
if __name__ == "__main__":
    main()